src/frontend/js/data/useCourseSearchParams/useCourseSearchParams.ts

```diff
@@ -0,0 +1,85 @@
+import { parse, stringify } from 'query-string';
+import { createContext, useEffect, useReducer } from 'react';
+
+import { APIListRequestParams } from '../../types/api';
+import { FilterDefinitionWithValues } from '../../types/filters';
+import { computeNewFilterValue } from '../../utils/filters/computeNewFilterValue';
+import { history, location } from '../../utils/indirection/window';
+
+interface FilterAction {
+  filter: FilterDefinitionWithValues;
+  payload: string;
+  type: 'FILTER_ADD' | 'FILTER_REMOVE';
+}
+
+interface QueryAction {
+  query: string;
+  type: 'QUERY_UPDATE';
+}
+
+export type CourseSearchParamsReducerAction = FilterAction | QueryAction;
+
+type CourseSearchParamsState = [
+  APIListRequestParams,
+  React.Dispatch<CourseSearchParamsReducerAction>
+];
+
+export const CourseSearchParamsContext = createContext<CourseSearchParamsState>(
+  [] as any,
+);
+
+const courseSearchParamsReducer = (
+  courseSearchParams: APIListRequestParams,
+  action: CourseSearchParamsReducerAction,
+) => {
+  switch (action.type) {
+    case 'QUERY_UPDATE':
+      return {
+        ...courseSearchParams,
+        // By replacing the empty string (the only falsy value we could receive for action.query) with `undefined`,
+        // we keep a clean interface and ensure `stringify` removes `&query=` from the query string in history.
+        query: action.query || undefined,
+      };
+
+    case 'FILTER_ADD':
+    case 'FILTER_REMOVE':
+      return {
+        ...courseSearchParams,
+        [action.filter.name]: computeNewFilterValue(
+          courseSearchParams[action.filter.name],
+          {
+            action: (action.type === 'FILTER_ADD' && 'add') || 'remove',
+            isDrilldown: !!action.filter.is_drilldown,
+            payload: action.payload,
+          },
+        ),
+      };
+  }
```

Comment:
You should maybe add a default case ? I know It should never happen but...

Reply:
I think if we add a default then we'll no longer be warned if we add new action types and forget to handle them in the reducer (as they'll be considered handled by the default).

At the same time action types are hardcoded at all the call sites (and probably not something we'd need manual typecasting to manage), so we can be pretty sure typescript will not let us dispatch something that is not one of the existing three. (and will force us to update everything if we add a new one)

Reply:
If you want an error, throw an error  after the switch. This code should never be reached.

Reply:
I'd rather have a compile-time warning than an error. In my opinion this is one instance where we can just rely on the compiler